File: app/src/main/java/app/quranhub/ui/mushaf/presenter/Mus7fPresenterImp.java
package app.quranhub.ui.mushaf.presenter;

import android.content.Context;

import androidx.annotation.NonNull;

import java.util.ArrayList;

import app.quranhub.data.Constants;
import app.quranhub.R;
import app.quranhub.ui.base.BasePresenterImp;
import app.quranhub.ui.mushaf.data.entity.Aya;
import app.quranhub.ui.mushaf.data.entity.TranslationBook;
import app.quranhub.ui.mushaf.interactor.Mus7fInteractor;
import app.quranhub.ui.mushaf.interactor.Mus7fInteractorImp;
import app.quranhub.ui.mushaf.model.QuranPageInfo;
import app.quranhub.ui.mushaf.model.SuraVersesNumber;
import app.quranhub.ui.mushaf.view.MushafView;
import app.quranhub.util.AppPreferencesUtils;

public class Mus7fPresenterImp extends BasePresenterImp<MushafView>
        implements Mus7fPresenter<MushafView>, Mus7fInteractor.ResultListener {

    private static final String TAG = Mus7fPresenterImp.class.getSimpleName();

    @NonNull
    private Mus7fInteractor interactor;
    @NonNull
    private Context context;

    public Mus7fPresenterImp(@NonNull Context context) {
        this.context = context;
        this.interactor = new Mus7fInteractorImp(this, context);
    }

    @Override
    public void getQuranPageInfo(int currentPage) {
        currentPage = Constants.Quran.NUM_OF_PAGES - currentPage;
        interactor.getPageInfo(currentPage);
    }

    @Override
    public void setNightMode(boolean nightMode) {
        AppPreferencesUtils.persistNightModeSetting(context, nightMode);
    }

    @Override
    public boolean getNightMode() {
        return AppPreferencesUtils.getNightModeSetting(context);
    }

    @Override
    public boolean toggleNightMode() {
        boolean nightMode = !getNightMode();
        setNightMode(nightMode);
        return nightMode;
    }


    @Override
    public void getAyaTafseer(int ayaId) {
        interactor.getAyaTafseer(ayaId);
    }

    @Override
    public void getCurrentTafseerBook(String currentTafsserId) {
        interactor.getTafseerBook(currentTafsserId);
    }

    @Override
    public void onGetPageInfo(QuranPageInfo pageInfo) {
        if (isViewAttached()) {
            baseView.showQuranPageInfo(pageInfo);
        }
    }

    @Override
    public void onGetAyaTafseer(String tafseer) {
        if (isViewAttached()) {
            baseView.onGetAyaTafseer(tafseer);
        }
    }

    @Override
    public void onGetTafsserBook(TranslationBook book) {
        if (isViewAttached()) {
            interactor.initTranslationDB(book.getDatabaseName());
            baseView.onGetTafseerBook(book);
        }
    }

    @Override
    public void getSurasInPage() {
        interactor.getPageSuras();
    }


    @Override
    public void onGetSuraPage(ArrayList<ArrayList<Integer>> suras) {
        if (isViewAttached()) {
            baseView.onGetPageSuras(suras);
        }
    }

    @Override
    public void onErroOccured() {
        if (isViewAttached()) {
            baseView.showMessage(context.getString(R.string.page_info_failed));
        }
    }

    @Override
    public void checkAyaHasRecorder(int id) {
        interactor.checkAyaHasRecorder(id);
    }

    @Override
    public void saveRecorderPath(int ayaId, String recorderPath) {
        if (isViewAttached()) {
            interactor.saveRecorderPath(ayaId, recorderPath);
        }
    }

    @Override
    public void deleteAyaVoiceRecorder(int ayaId) {
        if (isViewAttached()) {
            interactor.deleteAyaVoiceRecorder(ayaId);
        }
    }

    @Override
    public void getSuraNumofVerses() {
        if (isViewAttached()) {
            interactor.getSuraNumofVerses();
        }
    }

    @Override
    public void getFromAyaPage(int fromAya) {
        if (isViewAttached()) {
            interactor.getFromAyaPage(fromAya);
        }
    }

    @Override
    public void getNotificationAya(int ayaId) {
        if (isViewAttached()) {
            interactor.getAya(ayaId);
        }
    }

    @Override
    public void onNoBooks() {
        baseView.onNoBooksExist();
    }

    @Override
    public void onAyaHasRecorder(String recorderPath) {
        if (isViewAttached()) {
            baseView.onGetAyaRecorder(recorderPath);
        }
    }

    @Override
    public void onGetSuraVersesNumber(ArrayList<SuraVersesNumber> suraVersesNumbers) {
        if (isViewAttached()) {
            baseView.onGetSuraVersesNumber(suraVersesNumbers);
        }
    }

    @Override
    public void onGetAyaPage(int page) {
        if (isViewAttached()) {
            baseView.onGetAyaPage(page);
        }
    }

    @Override
    public void onGetAya(Aya aya) {
        if (isViewAttached()) {
            baseView.onGetCurrentAyaFromNotification(aya);
        }
    }


}

File: app/src/main/java/app/quranhub/ui/mushaf/presenter/Mus7fPresenterImp.kt
package app.quranhub.ui.mushaf.presenter

import android.content.Context
import app.quranhub.R
import app.quranhub.data.Constants
import app.quranhub.data.local.entity.Aya
import app.quranhub.data.local.entity.TranslationBook
import app.quranhub.data.local.prefs.AppPreferencesManager
import app.quranhub.ui.base.BasePresenterImp
import app.quranhub.ui.mushaf.interactor.Mus7fInteractor
import app.quranhub.ui.mushaf.interactor.Mus7fInteractorImp
import app.quranhub.ui.mushaf.model.QuranPageInfo
import app.quranhub.ui.mushaf.model.SuraVersesNumber
import app.quranhub.ui.mushaf.view.MushafView

class Mus7fPresenterImp(context: Context) : BasePresenterImp<MushafView>(),
    Mus7fPresenter<MushafView>, Mus7fInteractor.ResultListener {

    private val interactor: Mus7fInteractor = Mus7fInteractorImp(this, context)
    private val context: Context = context

    override fun getQuranPageInfo(currentPage: Int) {
        val adjustedPage = Constants.Quran.NUM_OF_PAGES - currentPage
        interactor.getPageInfo(adjustedPage)
    }

    override fun setNightMode(nightMode: Boolean) {
        AppPreferencesManager.persistNightModeSetting(context, nightMode)
    }

    override val nightMode: Boolean
        get() = AppPreferencesManager.getNightModeSetting(context)

    override fun toggleNightMode(): Boolean {
        val newNightMode = !nightMode
        setNightMode(newNightMode)
        return newNightMode
    }

    override val quranPageZoomScaleFactor: Float
        get() = AppPreferencesManager.getQuranPageZoomScaleSetting(context)

    override fun getAyaTafseer(ayaId: Int) {
        interactor.getAyaTafseer(ayaId)
    }

    override fun getCurrentTafseerBook(currentTafsserId: String) {
        interactor.getTafseerBook(currentTafsserId)
    }

    override fun onGetPageInfo(pageInfo: QuranPageInfo) {
        if (isViewAttached) {
            baseView!!.showQuranPageInfo(pageInfo)
        }
    }

    override fun onGetAyaTafseer(tafseer: String) {
        if (isViewAttached) {
            baseView!!.onGetAyaTafseer(tafseer)
        }
    }

    override fun onGetTafsserBook(book: TranslationBook) {
        if (isViewAttached) {
            interactor.initTranslationDB(book.databaseName)
            baseView!!.onGetTafseerBook(book)
        }
    }

    override fun getSurasInPage() {
        interactor.getPageSuras()
    }

    override fun onGetSuraPage(suras: ArrayList<ArrayList<Int>>) {
        if (isViewAttached) {
            baseView!!.onGetPageSuras(suras)
        }
    }

    override fun onErrorOccurred() {
        if (isViewAttached) {
            baseView!!.showMessage(context.getString(R.string.page_info_failed))
        }
    }

    override fun checkAyaHasRecorder(id: Int) {
        interactor.checkAyaHasRecorder(id)
    }

    override fun saveRecorderPath(ayaId: Int, recorderPath: String) {
        if (isViewAttached) {
            interactor.saveRecorderPath(ayaId, recorderPath)
        }
    }

    override fun deleteAyaVoiceRecorder(ayaId: Int) {
        if (isViewAttached) {
            interactor.deleteAyaVoiceRecorder(ayaId)
        }
    }

    override fun getSuraNumofVerses() {
        if (isViewAttached) {
            interactor.getSuraNumofVerses()
        }
    }

    override fun getFromAyaPage(fromAya: Int) {
        if (isViewAttached) {
            interactor.getFromAyaPage(fromAya)
        }
    }

    override fun getNotificationAya(ayaId: Int) {
        if (isViewAttached) {
            interactor.getAya(ayaId)
        }
    }

    override fun onNoBooks() {
        baseView!!.onNoBooksExist()
    }

    override fun onAyaHasRecorder(recorderPath: String) {
        if (isViewAttached) {
            baseView!!.onGetAyaRecorder(recorderPath)
        }
    }

    override fun onGetSuraVersesNumber(suraVersesNumbers: ArrayList<SuraVersesNumber>) {
        if (isViewAttached) {
            baseView!!.onGetSuraVersesNumber(suraVersesNumbers)
        }
    }

    override fun onGetAyaPage(page: Int) {
        if (isViewAttached) {
            baseView!!.onGetAyaPage(page)
        }
    }

    override fun onGetAya(aya: Aya) {
        if (isViewAttached) {
            baseView!!.onGetCurrentAyaFromNotification(aya)
        }
    }
}
